File: Modules/module_jetpack/src/main/java/com/example/william/my/module/jetpack/activity/PagingActivity.java
package com.example.william.my.module.jetpack.activity;

import android.os.Bundle;
import android.util.Log;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.paging.CombinedLoadStates;
import androidx.paging.PagingData;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.example.william.my.library.base.BaseActivity;
import com.example.william.my.module.bean.ArticleBean;
import com.example.william.my.module.jetpack.R;
import com.example.william.my.module.jetpack.adapter.PagingAdapter;
import com.example.william.my.module.jetpack.comparator.ArticleComparator;
import com.example.william.my.module.jetpack.model.ArticleViewModel;
import com.example.william.my.module.router.ARouterPath;

import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.functions.Consumer;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

import static autodispose2.AutoDispose.autoDisposable;

/**
 * https://developer.android.google.cn/topic/libraries/architecture/paging/v3-overview
 */
@Route(path = ARouterPath.JetPack.JetPack_Paging)
public class PagingActivity extends BaseActivity {

    private PagingAdapter mPagingAdapter;

    private final CompositeDisposable mDisposable = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.jet_activity_paging);

        mPagingAdapter = new PagingAdapter(new ArticleComparator());
        RecyclerView mRecycleView = findViewById(R.id.paging_recycleView);
        mRecycleView.setLayoutManager(new LinearLayoutManager(this));
        mRecycleView.setAdapter(mPagingAdapter);

        ArticleViewModel mViewModel = new ViewModelProvider(this).get(ArticleViewModel.class);

        //initArticleLiveData(mViewModel);

        initArticleFlowable(mViewModel);

        mPagingAdapter.addLoadStateListener(new Function1<CombinedLoadStates, Unit>() {
            @Override
            public Unit invoke(CombinedLoadStates combinedLoadStates) {
                Log.e("TAG", combinedLoadStates.getRefresh().toString());
                return null;
            }
        });
    }

    private void initArticleLiveData(ArticleViewModel viewModel) {
        viewModel.getArticleLiveData().observe(this, new Observer<PagingData<ArticleBean.DataBean.ArticleDetailBean>>() {
            @Override
            public void onChanged(PagingData<ArticleBean.DataBean.ArticleDetailBean> pagingData) {
                mPagingAdapter.submitData(getLifecycle(), pagingData);
            }
        });
    }

    private void initArticleFlowable(ArticleViewModel viewModel) {
        mDisposable.add(viewModel.getArticleFlowable()
                .to(autoDisposable(AndroidLifecycleScopeProvider.from(this)))
                .subscribe(new Consumer<PagingData<ArticleBean.DataBean.ArticleDetailBean>>() {
                    @Override
                    public void accept(PagingData<ArticleBean.DataBean.ArticleDetailBean> pagingData) throws Throwable {
                        mPagingAdapter.submitData(getLifecycle(), pagingData);
                    }
                }));
    }

    @Override
    protected void onStop() {
        super.onStop();
        // clear all the subscriptions
        mDisposable.clear();
    }
}

File: Modules/module_kotlin/src/main/kotlin/com/example/william/my/module/kotlin/activity/PagingActivity.kt
package com.example.william.my.module.kotlin.activity

import android.os.Bundle
import android.util.Log
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.paging.PagingData
import androidx.recyclerview.widget.LinearLayoutManager
import com.alibaba.android.arouter.facade.annotation.Route
import com.example.william.my.library.base.BaseActivity
import com.example.william.my.module.bean.ArticleBean
import com.example.william.my.module.kotlin.adapter.ArticleAdapter
import com.example.william.my.module.kotlin.comparator.ArticleComparator
import com.example.william.my.module.kotlin.databinding.KActivityPagingBinding
import com.example.william.my.module.kotlin.holder.ExampleLoadStateAdapter
import com.example.william.my.module.kotlin.model.PagingViewModel
import com.example.william.my.module.router.ARouterPath
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

/**
 * Paging
 * https://developer.android.google.cn/topic/libraries/architecture/paging/v3-overview
 */
@Route(path = ARouterPath.Kotlin.Kotlin_Paging)
class PagingActivity : BaseActivity() {

    private val mDisposable = CompositeDisposable()

    lateinit var binding: KActivityPagingBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = KActivityPagingBinding.inflate(layoutInflater)
        setContentView(binding.root)

        //val viewModel by viewModels<ExampleViewModel>()
        val viewModel = ViewModelProvider(this).get(PagingViewModel::class.java)

        val pagingAdapter = ArticleAdapter(ArticleComparator())
        val recycleView = binding.pagingRecycleView
        recycleView.layoutManager = LinearLayoutManager(this)
        recycleView.adapter = pagingAdapter

        //initArticleLiveData(viewModel, pagingAdapter)

        initArticleFlowable(viewModel, pagingAdapter)

        //获取加载状态
        pagingAdapter.addLoadStateListener {
            when (it.refresh) {
                is LoadState.NotLoading -> Log.e("TAG", "is NotLoading")
                is LoadState.Loading -> Log.e("TAG", "is Loading")
                is LoadState.Error -> Log.e("TAG", "is Error")
            }
        }

        //呈现加载状态
        //需要把withLoadStateFooter返回的adapter设置给recyclerview
        recycleView.adapter = pagingAdapter.withLoadStateHeaderAndFooter(
            header = ExampleLoadStateAdapter(pagingAdapter::retry),
            footer = ExampleLoadStateAdapter(pagingAdapter::retry)
        )
    }

    override fun onStop() {
        super.onStop()

        // clear all the subscriptions
        mDisposable.clear()
    }

    /**
     * Paging Coroutines -> LiveData
     */
    private fun initArticleLiveData(viewModel: PagingViewModel, pagingAdapter: ArticleAdapter) {
        // activity 可以使用 lifecycleScope。fragment 需要使用 viewLifecycleOwner.lifecycleScope
        // Activities can use lifecycleScope directly, but Fragments should instead use viewLifecycleOwner.lifecycleScope.
        lifecycleScope.launch {
            viewModel.articleFlow.collectLatest { pagingData ->
                pagingAdapter.submitData(pagingData)
            }
        }
    }

    /**
     * Paging RxJava -> Flowable
     */
    private fun initArticleFlowable(viewModel: PagingViewModel, pagingAdapter: ArticleAdapter) {
        mDisposable.add(
            viewModel.articleFlowable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(Consumer<PagingData<ArticleBean.DataBean.ArticleDetailBean>> {
                    pagingAdapter.submitData(lifecycle, it)
                })
        )
    }
}